### PR TITLE
Remove hardcoded "rhpkg".

### PR DESCRIPTION
**Pull request by junaruga**
This fixes https://github.com/sclorg/rpm-list-builder/issues/66 and https://github.com/sclorg/rpm-list-builder/issues/83 .

* Added --pkg-cmd PKG_CMD command option
  PKG_CMD works in build type: mock, copr.
* Added download type: fedpkg

Specification is here https://github.com/sclorg/rpm-list-builder/issues/83#issue-245364831 .

### CODE REVIEW

**Comment by junaruga** [comment]
I tested below cases too as acceptance test just in case.

```
$ rpmlb \
  --download fedpkg \
  --build mock \
  --mock-config fedora-rawhide-x86_64 \
  --branch master \
  --verbose \
  tests/fixtures/recipes/joke.yml \
  joke 2>&1 | tee rpmlb.log
```

```  
$ rpmlb \
  --download fedpkg \
  --build mock \
  --pkg-cmd fedpkg \
  --mock-config fedora-rawhide-x86_64 \
  --branch master \
  --verbose \
  tests/fixtures/recipes/joke.yml \
  joke 2>&1 | tee rpmlb.log
```

```
$ rpmlb \
  --download rhpkg \
  --build mock \
  --pkg-cmd rhpkg \
  --mock-config rhscl-2.4-rh-ror50-rhel-7-x86-64 \
  --branch rhscl-2.4-rh-ror50-rhel-7 \
  --verbose \
  tests/fixtures/recipes/ror.yml \
  rh-ror50 2>&1 | tee rpmlb.log
```

```
$ rpmlb \
  --download fedpkg \
  --build copr \
  --copr-repo rh-ror50-test \
  --branch master \
  --verbose \
  tests/fixtures/recipes/joke.yml \
  joke 2>&1 | tee rpmlb.log
```

```
$ rpmlb \
  --download rhpkg \
  --build mock \
  --mock-config rhscl-2.4-rh-ror50-rhel-7-x86-64 \
  --branch rhscl-2.4-rh-ror50-rhel-7 \
  --verbose \
  tests/fixtures/recipes/ror.yml \
  rh-ror50 2>&1 | tee rpmlb.log
```

File: tests/test_cli.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 import pytest
 from click.testing import CliRunner
 
-from rpmlb.cli import run
+from rpmlb import cli
```
Comment (junaruga) [inline tests/test_cli.py]: This line is to use 2 methods in rpmlb.cli module in `test_cli.py` .

The way is from
http://docs.python-guide.org/en/latest/writing/structure/
> Best
> x = modu.sqrt(4)  # sqrt is visibly part of modu's namespace

File: rpmlb/downloader/base_rpkg.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from .. import utils
+from .base import BaseDownloader
+
+LOG = logging.getLogger(__name__)
+
+
+class BaseRpkgDownloader(BaseDownloader):
+    """A downloader class to get a pacakge with package command."""
+
+    def command(self):
```
Comment (khardix) [inline rpmlb/downloader/base_rpkg.py]: This should be a `@property`.

I would envision the whole inheritance tree to look something like this:

```python
class Base:
    @property
    def command(self):
        raise NotImplementedError('...')

    def echo_command(self):
         """Example method that uses the command property."""
        print(self.command)


class A(Base):
    command = 'rhpkg'


class B(Base):
    command = 'fedpkg'
```
```python
>>> a, b = A(), B()
>>> a.echo_command()
rhpkg
>>> b.echo_command()
fedpkg
```

File: rpmlb/downloader/fedpkg.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+from .base_rpkg import BaseRpkgDownloader
+
+
+class FedpkgDownloader(BaseRpkgDownloader):
+    def command(self):
```
Comment (khardix) [inline rpmlb/downloader/fedpkg.py]: See comment for `BaseRpkgDownloader.command`.

File: rpmlb/downloader/rhpkg.py [hunk 4]
```diff
@@ -1,35 +1,6 @@
...
-
-        LOG.debug('rhpkg co %s', package)
-        subprocess.check_call(['rhpkg', 'co', package])
-        LOG.debug('cd %s', package)
-        os.chdir(package)
-        LOG.debug('git checkout %s', branch)
-        subprocess.check_call(['git', 'checkout', branch])
+class RhpkgDownloader(BaseRpkgDownloader):
+    def command(self):
+        return 'rhpkg'
```
Comment (khardix) [inline rpmlb/downloader/rhpkg.py]: See comment for `BaseRpkgDownloader.command`.

File: rpmlb/cli.py [hunk 5]
```diff
@@ -126,3 +135,15 @@ def run(recipe_file, collection_id, **option_dict):
     builder.run(work, **option_dict)
 
     log.info('Success!')
+
+
+def _adjust_option_dict(option_dict):
```
Comment (khardix) [inline rpmlb/cli.py]: This is a very generic name. Lets call it `_choose_pkg_cmd` or something like that.

File: rpmlb/cli.py [hunk 6]
```diff
@@ -126,3 +135,15 @@ def run(recipe_file, collection_id, **option_dict):
...
+    if not option_dict:
+        return
+    download = option_dict.get('download', None)
+    # When specifying download type: fedpkg or rhpkg,
+    # users must want to use the command in Builder.
```
Comment (khardix) [inline rpmlb/cli.py]: That is not true. What about rebuilding from Fedora to (EP)EL/CentOS? In that case you might want a `fedpkg` download, but `rhpkg` build.

Better solution would be setting the `pkg_cmd` default to `None` and do something like this:

```python
# If the user did not request a pkg command,
# choose one based on relevant download type.

# user chose the command explicitly
if option_dict['pkg_cmd'] is not None:
    return

# no explicit command, but relevant builder was selected
if download in ('fedpkg', 'rhpkg'):
    option_dict['pkg_cmd'] = download
    return

# no pkg_cmd or relevant builder – set default value
option_dict['pkg_cmd'] = 'fedpkg'
```

At that point, the functions starts to look as a prime candidate for `pkg_cmd` callback, although that might be more complicated to use.

File: rpmlb/downloader/base_rpkg.py (hunk 2 above)
Comment (hroncok) [inline rpmlb/downloader/base_rpkg.py]: Definitively agreed (commenting, so you get e-mail).

Comment (junaruga) [inline rpmlb/downloader/base_rpkg.py]: I also agree :)

Comment (junaruga) [inline rpmlb/downloader/base_rpkg.py]: I could not have the idea last time.

File: rpmlb/downloader/fedpkg.py (hunk 3 above)
Comment (junaruga) [inline rpmlb/downloader/fedpkg.py]: Yes, ok!

File: rpmlb/downloader/rhpkg.py (hunk 4 above)
Comment (junaruga) [inline rpmlb/downloader/rhpkg.py]: Yes, ok!

File: rpmlb/cli.py (hunk 5 above)
Comment (junaruga) [inline rpmlb/cli.py]: ok, the concrete name makes sense. I will fix it.

File: rpmlb/cli.py (hunk 6 above)
Comment (junaruga) [inline rpmlb/cli.py]: OK, it looks better. I will fix it.

Maybe below looks better.
Ref: http://docs.python-guide.org/en/latest/writing/style/#check-if-variable-equals-a-constant

```
# user chose the command explicitly
if not option_dict['pkg_cmd']:
    return
```

Comment (khardix) [inline rpmlb/cli.py]: I tend to be explicit with `None`, as for me it signifies "unknown or unset value" in opposite to "set value that happens to be falsey", (i.e. empty string). But this boils down to personal preference, so I'm fine with your proposed solution too.

Comment (junaruga) [inline rpmlb/cli.py]: Ah okay I could understand your intention.
However the comment "user chose the command explicitly" helps users to understand your intention.
So, I will keep this code.
Thanks for explanation.

**Comment by junaruga** [comment]
Merged with squash.
The document is later.

**Pull request merged**